tests/FSharp.Test.Utilities/Compiler.fs

```diff
@@ -0,0 +1,497 @@
+            assertErrorsLength source expected
+
+        let private assertErrors (what: string) libAdjust (source: ErrorInfo list) (expected: ErrorInfo list) : unit =
+            let errors = source |> List.map (fun error -> { error with Range = adjustRange error.Range libAdjust })
+
+            let inline checkEqual k a b =
+             if a <> b then
+                 Assert.AreEqual(a, b, sprintf "%s: Mismatch in %s, expected '%A', got '%A'.\nAll errors:\n%A" what k a b errors)
+
+            // TODO: Check all "categories", collect all results and print alltogether.
+            checkEqual "Errors count"  expected.Length errors.Length
+
+            List.zip errors expected
+            |> List.iter (fun (actualError, expectedError) ->
+                           let { Error = actualError; Range = actualRange; Message = actualMessage } = actualError
+                           let { Error = expectedError; Range = expectedRange; Message = expectedMessage } = expectedError
+                           checkEqual "Error" expectedError actualError
+                           checkEqual "ErrorRange" expectedRange actualRange
+                           checkEqual "Message" expectedMessage actualMessage)
+            ()
+
+        let adjust (adjust: int) (result: CompilationResult) : CompilationResult =
+            match result with
+            | Success s -> Success { s with Adjust = adjust }
+            | Failure f -> Failure { f with Adjust = adjust }
+
+        let shouldSucceed (result: CompilationResult) : CompilationResult =
+            match result with
+            | Success _ -> result
+            | Failure r ->
+                let message = sprintf "Compilation failed (expected to succeed).\n All errors:\n%A" (r.Errors @ r.Warnings)
+                failwith message
+
+        let shouldFail (result: CompilationResult) : CompilationResult =
+            match result with
+            | Success _ -> failwith "Compilation was \"Success\" (expected: \"Failure\")."
+            | Failure _ -> result
+
+        let private assertResultsCategory (what: string) (selector: Output -> ErrorInfo list) (expected: ErrorInfo list) (result: CompilationResult) : CompilationResult =
+            match result with
+             | Success r | Failure r ->
+                assertErrors what r.Adjust (selector r) expected
+            result
+
+        let withResults (expectedResults: ErrorInfo list) result : CompilationResult =
+            assertResultsCategory "Results" (fun r -> r.Warnings @ r.Errors) expectedResults result
+
+        let withResult (expectedResult: ErrorInfo ) (result: CompilationResult) : CompilationResult =
+            withResults [expectedResult] result
+
+        let with' (expected: (ErrorType * Line * Col * Line * Col * string) list) (result: CompilationResult) : CompilationResult =
```

Comment:
I think this should be renamed to something else. `withErrors` perhaps? or `withDiagnostics`? The `'` at the end is fine for locals defined in the body of an expression but not for a public API

Reply:
Yeah, I wasn't sure how to name it, hoped for some suggestions :)